### ISSUE TITLE
Port 2.0: Fix Unix missing file state

### ISSUE THREAD

**Issue by karelz** [dotnet/corefx#20040]
We need to port #19959

cc @JeremyKuhne @stephentoub

**Comment by danmosemsft** [comment]
Fix https://github.com/dotnet/corefx/issues/20034 first

**Comment by stephentoub** [comment]
@JeremyKuhne, please include the commits from https://github.com/dotnet/corefx/pull/20049 if/when this is ported to 2.0.

**Comment by JeremyKuhne** [comment]
https://github.com/dotnet/coreclr/pull/11757 is part of this as well. Root issue is #19850.

**Comment by CIPop** [comment]
Reopening:

Failing on Ubuntu:
https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_netcoreapp_ubuntu16.04_debug_prtest/23/consoleFull

```
System.IO.Tests.DirectoryInfo_Delete.DeleteRoot
System.IO.Tests.DirectoryInfo_Delete_bool.DeleteRoot
System.IO.Tests.Directory_Delete_str_bool.DeleteRoot
System.IO.Tests.Directory_Delete_str.DeleteRoot
```

https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_netcoreapp_ubuntu16.04_debug_prtest/23/testReport/junit/System.IO.Tests/Directory_Delete_str/DeleteRoot/
```
Regression

System.IO.Tests.Directory_Delete_str.DeleteRoot (from (empty))

Failing for the past 1 build (Since Failed#23 )
Took 0 ms.
add description
Stacktrace

MESSAGE:
Assert.Throws() Failure\nExpected: typeof(System.IO.IOException)\nActual: typeof(System.IO.DirectoryNotFoundException): Could not find a part of the path '/'.
+++++++++++++++++++
STACK TRACE:
at System.IO.UnixFileSystem.RemoveDirectory(String fullPath, Boolean recursive) in /mnt/j/workspace/dotnet_corefx/master/outerloop_netcoreapp_ubuntu16.04_debug_prtest/src/System.IO.FileSystem/src/System/IO/UnixFileSystem.cs:line 317 at System.IO.Directory.Delete(String path) in /mnt/j/workspace/dotnet_corefx/master/outerloop_netcoreapp_ubuntu16.04_debug_prtest/src/System.IO.FileSystem/src/System/IO/Directory.cs:line 582 at System.IO.Tests.Directory_Delete_str.Delete(String path) in /mnt/j/workspace/dotnet_corefx/master/outerloop_netcoreapp_ubuntu16.04_debug_prtest/src/System.IO.FileSystem/tests/Directory/Delete.cs:line 15 at System.IO.Tests.Directory_Delete_str.<DeleteRoot>b__6_0() in /mnt/j/workspace/dotnet_corefx/master/outerloop_netcoreapp_ubuntu16.04_debug_prtest/src/System.IO.FileSystem/tests/Directory/Delete.cs:line 67
```

https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_netcoreapp_ubuntu16.04_debug_prtest/23/testReport/junit/System.IO.Tests/Directory_Delete_str_bool/DeleteRoot/
```
System.IO.Tests.Directory_Delete_str_bool.DeleteRoot (from (empty))

Failing for the past 1 build (Since Failed#23 )
Took 1 ms.
add description
Stacktrace

MESSAGE:
Assert.Throws() Failure\nExpected: typeof(System.IO.IOException)\nActual: typeof(System.IO.DirectoryNotFoundException): Could not find a part of the path '/'.
+++++++++++++++++++
STACK TRACE:
at System.IO.UnixFileSystem.RemoveDirectory(String fullPath, Boolean recursive) in /mnt/j/workspace/dotnet_corefx/master/outerloop_netcoreapp_ubuntu16.04_debug_prtest/src/System.IO.FileSystem/src/System/IO/UnixFileSystem.cs:line 317 at System.IO.Directory.Delete(String path, Boolean recursive) in /mnt/j/workspace/dotnet_corefx/master/outerloop_netcoreapp_ubuntu16.04_debug_prtest/src/System.IO.FileSystem/src/System/IO/Directory.cs:line 589 at System.IO.Tests.Directory_Delete_str_bool.Delete(String path) in /mnt/j/workspace/dotnet_corefx/master/outerloop_netcoreapp_ubuntu16.04_debug_prtest/src/System.IO.FileSystem/tests/Directory/Delete.cs:line 216 at System.IO.Tests.Directory_Delete_str.<DeleteRoot>b__6_0() in /mnt/j/workspace/dotnet_corefx/master/outerloop_netcoreapp_ubuntu16.04_debug_prtest/src/System.IO.FileSystem/tests/Directory/Delete.cs:line 67
```

**Comment by CIPop** [comment]
Also on OSX: https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_netcoreapp_osx_debug_prtest/39/consoleFull#9832431179fe3b83-f408-404c-b9e7-9207d232e5fc
(same errors)

**Comment by JeremyKuhne** [comment]
@CIPop I assume you meant to re-open #20034?

**Comment by CIPop** [comment]
The last comments weren't clear but I assumed we wanted to track issues via this bug.
Copying comments there and reopening.

**Comment by JeremyKuhne** [comment]
This has been fixed and ported to 2.0.
